be/src/exec/pipeline/crossjoin/cross_join_left_operator.cpp

```diff
@@ -0,0 +1,419 @@
+                        row_count = _probe_chunk_index - _probe_rows_index;
+                    }
+
+                    _copy_joined_rows_with_index_base_build(chunk, row_count, _probe_rows_index,
+                                                            _beyond_threshold_build_rows_index);
+                    _probe_rows_index += row_count;
+
+                    if (_probe_rows_index == _probe_chunk_index) {
+                        ++_beyond_threshold_build_rows_index;
+                        _probe_rows_index = 0;
+                    }
+
+                    // _probe_chunk is done with _build_chunk.
+                    if (_beyond_threshold_build_rows_index >= _total_build_rows) {
+                        _probe_chunk = nullptr;
+                    }
+                } else {
+                    // if remain rows of right is bigger than left, we should scan right based on left.
+                    if (row_count > _total_build_rows - _beyond_threshold_build_rows_index) {
+                        row_count = _total_build_rows - _beyond_threshold_build_rows_index;
+                    }
+
+                    _copy_joined_rows_with_index_base_probe(chunk, row_count, _probe_rows_index,
+                                                            _beyond_threshold_build_rows_index);
+                    _beyond_threshold_build_rows_index += row_count;
+
+                    if (_beyond_threshold_build_rows_index == _total_build_rows) {
+                        ++_probe_rows_index;
+                        _beyond_threshold_build_rows_index = _build_rows_threshold;
+                    }
+
+                    // _probe_chunk is done with _build_chunk.
+                    if (_probe_rows_index >= _probe_chunk_index) {
+                        _probe_chunk = nullptr;
+                    }
+                }
+            } else if (_within_threshold_build_rows_index < _build_rows_threshold) {
+                // step 1:
+                // we scan all chunks of right table.
+                if (row_count > _build_rows_threshold - _within_threshold_build_rows_index) {
+                    row_count = _build_rows_threshold - _within_threshold_build_rows_index;
+                }
+
+                _copy_joined_rows_with_index_base_probe(chunk, row_count, _probe_chunk_index,
+                                                        _within_threshold_build_rows_index);
+                _within_threshold_build_rows_index += row_count;
+            } else {
+                // step policy decision:
+                DCHECK_EQ(_within_threshold_build_rows_index, _build_rows_threshold);
+
+                if (_build_rows_threshold != 0) {
+                    // scan right chunk_size rows for next row of left chunk.
+                    ++_probe_chunk_index;
+                    if (_probe_chunk_index < _probe_chunk->num_rows()) {
+                        _within_threshold_build_rows_index = 0;
```

Comment:
why reset _within_threshold_build_rows_index  here?

Reply:
Because for every row of left chunk, we should iterate right table from start.